### PR TITLE
Improve typescript typings with strict null check

### LINKED ISSUE #2588: Error at build
**maplibre-gl-js version**: 3.0.0

### Steps to Trigger Behavior

 1. Add maplibre-gl as a dependency
 2. `pnpm i`
 3. `pnpm run build`

Expected behavior: application compiles without error

Actual behavior:

```
Error: node_modules/.pnpm/maplibre-gl@3.0.0/node_modules/maplibre-gl/dist/maplibre-gl.d.ts:1145:2 - error TS2416: Property 'getDefault' in type 'BindVertexArray' is not assignable to the same property in base type 'BaseValue<WebGLVertexArrayObject>'.
  Type '() => WebGLVertexArrayObject | null' is not assignable to type '() => WebGLVertexArrayObject'.
    Type 'WebGLVertexArrayObject | null' is not assignable to type 'WebGLVertexArrayObject'.
      Type 'null' is not assignable to type 'WebGLVertexArrayObject'.

1145  getDefault(): WebGLVertexArrayObject | null;
```

I've been using version 3.0.0-pre.5 so far without encountering this error.

### PR DESCRIPTION
**Pull request by HarelM**
This is an initial commit.
Some things to take into consideration:
1. In most cases it's enough to add a `!` in order to solve the `undefined` part.
2. When deleting a property the property should be defined with `?`
3. Add typings to functions as much as possible, your future self will thank you.
4. When importing, use `import type` so that we only import the relevant needed type.

Generally speaking, there shouldn't be any logic changes.

You can use `npm run typecheck` to see a summary of the remaining issues.

I think PRs against this branch would be the right way to move forward in terms of review

Resolves #2588

### CODE REVIEW

File: build/generate-style-code.ts [hunk 1]
```diff
@@ -24,7 +39,7 @@ function nativeType(property) {
         case 'string':
             return 'string';
         case 'enum':
-            return Object.keys(property.values).map(v => JSON.stringify(v)).join(' | ');
+            return Object.keys(property.values!).map(v => JSON.stringify(v)).join(' | ');
```
[the remark this answers is not in the document]
Comment (HarelM) [reply build/generate-style-code.ts:42]: Feel free to pitch in 😀 I'll be happy to review any PR sent against this branch.

**Comment by HarelM** [comment]
The more I play around this the less value I see in pushing this forward honestly.
I'm mainly adding `!` in places where the code is expected not to have null or undefined but the transpiler is saying that it can be undefined.
Littering the code with `!`, `| null`, `| undefined` is not a great outcome.
While this might uncover some bugs or badly defined types in most cases is only adds noise.
I'll see if there's a good way to make sure that the generated `maplibre-gl.d.ts` file is using strict mode and add this to the CI instead of waisting a lot of energy in something that won't really help.
If anyone else has other opinions around this I'll be happy to hear.

**Comment by neodescis** [comment]
I agree with your assessment. I think that strictNullCheck is one of the least useful of the strict checks, especially in terms of introducing it to an existing project. We may want to look into the other strict checks though.

**Comment by drwestco** [comment]
It is difficult to adopt in a mature project, with years of loose type definitions to account for. Littering the code with `!` assertions is the wrong way to do it, though. Whether it's worth the effort to improve the actual type definitions is a judgment call. When turned on, `strictNullChecks` will correctly flag potential bugs.

```Typescript
        for (const feature of features) {
            const featureKey = feature.key;
            // Use of featureKey without null check flagged by compiler
            if (seenFeatures[featureKey.bucketInstanceId] === undefined) {
                seenFeatures[featureKey.bucketInstanceId] = {};
            }
            if (seenFeatures[featureKey.bucketInstanceId][featureKey.featureIndex]) {
                continue;
            }
```

If feature.key is guaranteed to be non-null, the type shouldn't indicate that it's optional. If it is optional, this code should check. The issue here is that a common `QueryResult` type is used for both `grid.query` and `grid.hitTest`, and `key` is only possibly null for the latter. That's sloppy. `grid.hitTest` should return a different type.

**Comment by HarelM** [comment]
Don't get me wrong, I'm all for improving the typings, but I find that the errors are not accurate in some cases, especially when the code is a bit more complicated like the following:
https://github.com/maplibre/maplibre-gl-js/blob/b45b5b40d83cfaaa791176b333afe0e1cd4c14f2/src/data/bucket/line_bucket.ts#L300

Adding some additional types for specific cases also isn't a great solution, defining TypeA and then TypeAButWithoutB just to satisfy this check doesn't feel in some cases like the right solution.
There are places, as you said which makes sense, but not all of them.

What I can think of as a middle ground is to enable this check, fix the places where types are missing, incorrect, or needs attention and improve the code just a bit as separate PRs and then decide what to do on the remaining places and whether or not to introduce this flag.

Having said that, making sure the public API is strict can be a lot easier, as I did in #2623.

**Comment by HarelM** [comment]
I'm not planning to follow this through.
If anyone else is interested, feel free.

**Pull request closed without merging**